//------------------------------------------------------------------------------
// {C} Copyright 2019 Pensando Systems Inc. All rights reserved
//
// protobuf specification for routing
//------------------------------------------------------------------------------

syntax = "proto3";
package tpc;

import public "gogo.proto";
import "meta/meta.proto";
import "types.proto";

service RouteSvc {
  rpc RouteTableCreate (RouteTableSpec) returns (RouteTableStatus) {}
}

// route object
message Route {
  types.IPPrefix  Prefix  = 1 [(gogoproto.nullable) = false];
  types.IPAddress NextHop = 2 [(gogoproto.moretags) = "meta:mandatory"];
  uint32          PCNId   = 3 [(gogoproto.moretags) = "meta:mandatory"];
}

// route table specification
// NOTE: any time a route table's route(s) changes, client is expected to send
//       full new route table (i.e., no incremental route add/del/upd is
//       supported)
message RouteTableSpec {
  // unique route table id (this id is not scoped under PCN or subnet)
  uint32         Id     = 1 [(gogoproto.moretags) = "meta:mandatory"];
  // IP address family
  types.IPAF     AF     = 2 [(gogoproto.moretags) = "meta:mandatory"];
  // routes in this route table
  repeated Route Routes = 3 [(gogoproto.nullable) = false];

}

// operational status of the route table, if any
message RouteTableStatus {
}

// route table object
message RouteTable {
  meta.TypeMeta     TypeMeta = 1 [(gogoproto.embed) = true, (gogoproto.nullable) = false, (gogoproto.jsontag) = ",inline"];
  meta.ObjMeta      ObjMeta  = 2 [(gogoproto.embed) = true, (gogoproto.nullable) = false, (gogoproto.jsontag) = "meta,omitempty"];
  RouteTableSpec    Spec     = 3 [(gogoproto.nullable) = false, (gogoproto.jsontag) = "spec,omitempty"];
  RouteTableStatus  Status   = 4 [(gogoproto.nullable) = false, (gogoproto.jsontag) = "status,omitempty"];
}
